airsonic-main/src/main/java/org/airsonic/player/service/InternetRadioService.java

```diff
@@ -130,19 +130,32 @@ public void clearInternetRadioSourceCache(Integer internetRadioId) {
 
     /**
      * Retrieve a list of sources from the given internet radio
+     * The radio stream can either by binary or a playlist pointing to binary streams
      *
      * This method uses a default maximum limit of PLAYLIST_REMOTE_MAX_LENGTH sources.
      *
      * @param radio an internet radio
      * @return a list of internet radio sources
      */
     private List<InternetRadioSource> retrieveInternetRadioSources(InternetRadio radio) throws Exception {
-        return retrieveInternetRadioSources(
-            radio,
-            PLAYLIST_REMOTE_MAX_LENGTH,
-            PLAYLIST_REMOTE_MAX_BYTE_SIZE,
-            PLAYLIST_REMOTE_MAX_REDIRECTS
-        );
+        String streamUrl = radio.getStreamUrl();
+        HttpURLConnection urlConnection = connectToURLWithRedirects(new URL(streamUrl), PLAYLIST_REMOTE_MAX_REDIRECTS);
+        String contentType = urlConnection.getContentType();
+        if ("audio/mpeg".equals(contentType)) {
+            //for direct binary streams
+            List<InternetRadioSource> entries = new ArrayList<>();
+            entries.add(new InternetRadioSource(streamUrl));
+            return entries;
+        } else {
+            //for playlists containing binary streams
+            return retrieveInternetRadioSources(
+                    radio,
+                    PLAYLIST_REMOTE_MAX_LENGTH,
+                    PLAYLIST_REMOTE_MAX_BYTE_SIZE,
+                    PLAYLIST_REMOTE_MAX_REDIRECTS
+            );
+        }
```

Comment:
I think this has 2 issues:
1. It's connecting twice. You connect to the URL in L142, but then will connect again inside `retrieveInternetRadioSources(...)`
2. There is a connection leak. The urlConnection in L142 is never disconnected or released.

I suggest
- collapse retrievePlaylist() and combine it with retrieveInternetRadioSources(radio, maxCount, ...) 
- do this business inside the try-finally of the (former) retrievePlaylist()

something like:
```
private List<InternetRadioSource> retrieveInternetRadioSources(InternetRadio radio, int maxCount, long maxByteSize, int maxRedirects) throws Exception {
        // Retrieve the remote playlist
        String playlistUrl = radio.getStreamUrl();
        LOG.debug("Parsing internet radio playlist at {}...", playlistUrl);
        SpecificPlaylist inputPlaylist = null;
        HttpURLConnection urlConnection = connectToURLWithRedirects(url, maxRedirects);
        try (InputStream in = urlConnection.getInputStream()) {
            String contentType = urlConnection.getContentType();
            if ("audio/mpeg".equals(contentType)) {
                //for direct binary streams
                return Collections.singletonList(new InternetRadioSource(streamUrl));
            }
            String contentEncoding = urlConnection.getContentEncoding();
            if (maxByteSize > 0) {
                inputPlaylist = SpecificPlaylistFactory.getInstance().readFrom(new BoundedInputStream(in, maxByteSize), contentEncoding);
            } else {
                inputPlaylist = SpecificPlaylistFactory.getInstance().readFrom(in, contentEncoding);
            }
        } finally {
            urlConnection.disconnect();
        }
        if (inputPlaylist == null) {
            throw new PlaylistFormatUnsupported("Unsupported playlist format " + url.toString());
        }
        // Retrieve stream URLs
        List<InternetRadioSource> entries = new ArrayList<>();
        ...
```

```